File: frontend/src/routes/games/TicTacToe.js
import {useEffect, useRef, useState} from "react";
import {useParams} from "react-router";

const emptyBoard = [[null, null, null], [null, null, null], [null, null, null]]

function TicTacToe(props) {
  const gameState = props.state?.board || emptyBoard
  const dispatchAction = props.dispatchAction;

  const play = (x, y) => {
    console.info('play', x, y)
    dispatchAction( {'@type': 'nl.hiddewieringa.game.tictactoe.PlaceMarkLocation', location: {x, y}})
  }

  return (<div>
    <p>TicTacToe</p>
    <table>
      <tbody>{
        [...Array(3)].map((_, y) => (
          <tr key={y}>{
            [...Array(3)].map((_, x) => (
              <td key={x} onClick={() => play(x, y)}>{gameState[x][y] !== null ? (gameState[x][y]['@type'] === 'nl.hiddewieringa.game.tictactoe.Cross' ? 'X' : 'O') : '.'}</td>
            ))
          }</tr>
        ))
      }</tbody>
    </table>
  </div>);
}

export default TicTacToe;

File: frontend/src/main/kotlin/nl/hiddewieringa/game/frontend/games/TicTacToe.kt
package nl.hiddewieringa.game.frontend.games

import kotlinx.css.*
import kotlinx.html.ButtonType
import kotlinx.html.js.onClickFunction
import nl.hiddewieringa.game.frontend.GameUiProps
import nl.hiddewieringa.game.frontend.ParametersProps
import nl.hiddewieringa.game.tictactoe.*
import nl.hiddewieringa.game.tictactoe.state.TicTacToePlayerState
import react.dom.*
import react.functionalComponent
import styled.css
import styled.styledP
import styled.styledTable
import styled.styledTd

val TicTacToeParametersComponent = functionalComponent<ParametersProps<TicTacToeGameParameters>> { props ->
    val startGame = props.startGame

    button(null, null, ButtonType.button, "uk-button uk-button-primary") {
        attrs.onClickFunction = { startGame(TicTacToeGameParameters.serializer(), TicTacToeGameParameters) }
        +"Start a new game"
    }
}

val emptyBoard = arrayOf(arrayOf<GameMark?>(null, null, null), arrayOf<GameMark?>(null, null, null), arrayOf<GameMark?>(null, null, null))

val TicTacToeComponent = functionalComponent<GameUiProps<TicTacToePlayerState, TicTacToePlayerActions>> { props ->
    val gameState = props.gameState

    val playerToPlay = gameState?.playerToPlay
    val board = gameState?.board ?: emptyBoard
    val playerWon = gameState?.playerWon
    val gameFinished = gameState?.gameFinished ?: false

    val dispatchAction = props.dispatchAction
    val playerId = props.playerId

    val play = { x: Int, y: Int ->
        console.info("play", x, y)
        val action = PlaceMarkLocation(Location(x, y))
        dispatchAction.dispatch(action)
    }

    div {
        styledP {
            css {
                textAlign = TextAlign.center
            }

            if (gameFinished) {
                +when {
                    playerWon == null -> "Nobody won: drow"
                    playerWon.toString() == playerId -> "You won!"
                    else -> "$playerWon won!"
                }
            } else if (playerToPlay != null) {
                if (playerToPlay.toString() == playerId) {
                    +"You play!"
                } else {
                    +"Current player: $playerToPlay"
                }
            }
        }

        if (!gameFinished) {
            styledTable {
                css {
                    width = 300.px
                    margin = "1rem auto"
                    border = "3px solid #333"
                    borderRadius = 3.px
                    borderCollapse = BorderCollapse.collapse
                }
                tbody {
                    (0 until 3).map { y ->
                        tr {
                            attrs.key = y.toString()
                            (0 until 3).map { x ->
                                styledTd {
                                    css {
                                        height = 100.px
                                        width = 100.px
                                        border = "3px solid #333"
                                        textAlign = TextAlign.center
                                        fontFamily = "sans-serif"
                                        fontSize = 48.px
                                    }
                                    attrs.key = x.toString()
                                    attrs.onClickFunction = { play(x, y) }

                                    +when {
                                        board[x][y] == null -> ""
                                        board[x][y] is Cross -> "×"
                                        else -> "⊙"
                                    }
                                }
                            }
                        }
                    }
                }
            }
        }
    }
}